components/cards/SessionCard.tsx

```diff
@@ -0,0 +1,271 @@
+  const { handlePress, item, handleBookMark } = props;
+  const { colors } = useTheme();
+
+  const timeSplitted: Array<string> = getSessionTime(item.start_time).split(' ');
+
+  return (
+    <TouchableWithoutFeedback testID="card-list" onPress={handlePress}>
+      <Row style={[styles.listCard, { backgroundColor: colors.card }]}>
+        <Row style={styles.sessionCardRow}>
+          <View style={styles.rowVerticalStart}>
+            <StyledText font="medium" size="lg">
+              {timeSplitted[0]}
+            </StyledText>
+            <StyledText font="medium" size="md" variant="text" style={styles.amOrPm}>
+              {timeSplitted[1]}
+            </StyledText>
+          </View>
+          <View style={styles.listCardDetails}>
+            <StyledText font="bold" size="lg">
+              {truncate(50, item.title)}
+            </StyledText>
+            <Space size={16} />
+            <StyledText font="regular" size="md" variant="text">
+              {truncate(72, item.description)}
+            </StyledText>
+            <Space size={13} />
+            <StyledText size="sm" font="light">
+              {getScheduleTimeAndLocation(item.start_date_time, item.end_date_time, item.rooms[0] as Room)}
+            </StyledText>
+            <Space size={13} />
+            {item.speakers.length > 0 && (
+              <Row style={styles.rowHorizontalStart}>
+                {item.speakers.map((speaker, index) => (
+                  <Row key={speaker.name}>
+                    <MaterialCommunityIcons name="android" size={20} color={colors.primary} />
+                    <Space size={10} horizontal />
+                    <StyledText font="regular" size="sm" variant="primary">
+                      {speaker.name}
+                    </StyledText>
+                    {index + 1 !== item.speakers.length && <Space size={12} horizontal />}
+                  </Row>
+                ))}
+              </Row>
+            )}
+          </View>
+        </Row>
+        <AntDesign
+          testID="bookmark"
+          name={item.is_bookmarked ? 'star' : 'staro'}
+          size={21}
+          color={item.is_bookmarked ? colors.tertiary : colors.primary}
+          onPress={handleBookMark}
+          style={styles.rowVerticalStart}
+        />
+      </Row>
+    </TouchableWithoutFeedback>
+  );
+};
+
+const SessionCard = ({
+  handlePress,
+  handleBookMark,
+  item,
+  screen = 'home',
+  variant = 'card',
+}: SessionCardProps<SessionForSchedule> | SessionCardProps<Session>) => {
+  if (screen === 'sessions' && variant === 'list') {
+    return (
+      <SessionCardList handlePress={handlePress} handleBookMark={handleBookMark} item={item as SessionForSchedule} />
+    );
+  } else if (screen === 'sessions') {
+    return <SessionCardOnSessions item={item as SessionForSchedule} handlePress={handlePress} />;
+  }
+  return <SessionCardOnHome item={item} handlePress={handlePress} />;
+};
```

Comment:
The generic `SessionCardProps<T>` type is problematic here which leads to the ` as ` assertions. We are not providing TypeScript with all the information we know about how this component should be used.

We are also not passing `handleBookmark` to `SessionCardOnSessions` and TypeScript did not throw an error because we marked it as optional in the type helper, but our application is broken since we cannot star the session from the card view.

I would suggest using a [Tagged Union Type (Discriminated Union)](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions) instead (there are different ways of writing it, here is one possible way):

```ts
type SessionCardHome = {
  screen: 'home';
  item: Session;
  handlePress: () => void;
};

type SessionCardSessions = {
  screen: 'sessions';
  item: SessionForSchedule;
  variant?: 'card' | 'list';
  handlePress: () => void;
  handleBookMark: () => void;
};

type SessionCardProps = SessionCardHome | SessionCardSessions;

const SessionCardOnHome = (props: Omit<SessionCardHome, 'screen'>) => {
  // <skipped>
}

const SessionCardOnSessions = (props: Omit<SessionCardSessions, 'screen'>) => {
  // <skipped>
}

const SessionCardList = (props: Omit<SessionCardSessions, 'screen'>) => {
  // <skipped>
}

const SessionCard = (props: SessionCardProps) => {
  if (props.screen === 'sessions') {
    const { handlePress, handleBookMark, item, variant = 'card' } = props;
    if (variant === 'list') {
      return <SessionCardList item={item} handleBookMark={handleBookMark} handlePress={handlePress} />;
    }
    return <SessionCardOnSessions item={item} handleBookMark={handleBookMark} handlePress={handlePress} />;
  }
  const { handlePress, item } = props;
  return <SessionCardOnHome item={item} handlePress={handlePress} />;
};
```

---

**Also:** We can use a tagged union type instead of `SessionListVerticalProps<T>` to provide TypeScript with more information.

Reply:
Nice!
Honestly, I didn't know a better way of organizing my types and I felt the whole component was messy. Thank you for the insight. I have learnt something new!